components/SignUp.js

```diff
@@ -1,81 +1,160 @@
-import styles from "../styles/signinup.module.css";
+import styles from "../styles/SignUp.module.css";
+import { useState, useEffect } from "react";
 
-export default function SignUp() {
+export default function signup() {
+  const initialValues = { name: "", phone: "", email: "", password: "", verifyPassword: "" };
+  const [formValues, setFormValues] = useState(initialValues);
+  const [formErrors, setFormErrors] = useState({});
+  const [passwordStrength, setpasswordStrength] = useState("Lösenordets styrka:");
+  const [isSubmit, setIsSubmit] = useState(false);
 
-  async function signUp() {
-    // When the user clicks the button, check that all fields are filled in, and if so, create a new user. Otherwise, return an alert.
-    if (
-      document.querySelector("#verify-password").value != document.querySelector("#password").value ||
-      document.querySelector("#name").value.length < 3 ||
-      document.querySelector("#phone").value.length < 10 ||
-      document.querySelector("#email").value.length < 6 ||
-      document.querySelector("#password").value.length < 6
-    ) {
-      return alert("Please fill in all fields correctly! (Name must be at least 3 characters, phone number must be at least 10 characters, email must be at least 6 characters, password must be at least 6 characters)");
-    } else {
-      const response = await fetch(
-        "http://localhost:3000/api/users/newuser",
-        {
-          method: "POST",
-          mode: "cors",
-          headers: { "Content-Type": "application/json" },
-          body: JSON.stringify({
-            name: document.querySelector("#name").value,
-            phone: document.querySelector("#phone").value,
-            email: document.querySelector("#email").value,
-            password: document.querySelector("#password").value,
-
-          }),
-        }
-      );
-      const data = await response.json();
-      if (data.message === "User created!") {
-        return alert("User created!");
-      } else if (data.message === "An account with that email is already in use!") {
-        return alert("An account with that email is already in use!");
-      } else {
-        return alert("Error creating user!");
-      }
-    }
-  }
+  // Handles input changes on submit.
+  const handleChange = (e) => {
+    const { name, value } = e.target;
+    setFormValues({ ...formValues, [name]: value });
+  };
 
-  // Password strength check.
-  function passwordStrength() {
-    const passwordInput = document.getElementById("password");
-    const StrengthIndicator =
-      document.querySelector("#strength-indicator");
+  // Handles submit.
+  const handleSubmit = (e) => {
+    e.preventDefault();
+    setFormErrors(validate(formValues));
+    setIsSubmit(true);
```

Comment:
The lifecycle here is a little bit complex, and unnecessarily so. Basically, this is what will happen with the code as it's written in this PR:

1. Component renders form
2. User enters data (for each change, the component is re-rendered)
3. User submits form
4. `handleSubmit()` is called
  a. `validate()` is called and returns an object
  b. The object is stored in the state referenced by `formErrors`
  c. The value `true` is stored in the state referenced by `isSubmit`
5. **Component re-renders (because state changed)**
6. Since `formErrors` changed, the callback passed to `useEffect` (line 25) will execute
  a. If the `formErrors` object is empty, data will be posted
  b. Otherwise, nothing happens until user tries to submit again, at which point `formErrors` will change and re-trigger step 6

Steps 5 and 6 are unnecessary. Instead, you could have just done something like this, here:

```js
const handleSubmit = (e) => {
  e.preventDefault();
  const errors = validate(formValues);

  // If there are no errors, submit the data
  if (Object.keys(errors).length === 0) {
    postData(formValues);
  } else {
    // There were errors! Store them in state to trigger a re-render
    // with the errors available to be displayed in the form UI
    setFormErrors(errors);
  }
};
```

This would remove the need for the `useEffect()` and also make the code less error-prone.